chain/chain/src/chain.rs

```diff
@@ -3722,30 +3742,38 @@ impl Chain {
     fn get_update_shard_job(
         &self,
         me: &Option<AccountId>,
-        block: &Block,
+        // TODO(#10584): introduce separate structure which can be derived from
+        // both Block and OptimisticBlock.
+        block_metadata: &BlockHeader,
+        chunk_headers: &Chunks,
+        shard_index: ShardIndex,
         prev_block: &Block,
-        chunk_header: &ShardChunkHeader,
         prev_chunk_header: &ShardChunkHeader,
-        shard_id: ShardId,
         mode: ApplyChunksMode,
-        incoming_receipts: &HashMap<ShardId, Vec<ReceiptProof>>,
+        incoming_receipts: Option<&Vec<ReceiptProof>>,
         storage_context: StorageContext,
     ) -> Result<Option<UpdateShardJob>, Error> {
         let _span = tracing::debug_span!(target: "chain", "get_update_shard_job").entered();
-        let prev_hash = block.header().prev_hash();
-        let shard_context = self.get_shard_context(me, block.header(), shard_id, mode)?;
+        let prev_hash = prev_block.hash();
+        let epoch_id = self.epoch_manager.get_epoch_id_from_prev_block(prev_hash)?;
+        let shard_layout = self.epoch_manager.get_shard_layout(&epoch_id)?;
+        let shard_id = shard_layout.get_shard_id(shard_index)?;
+        let chunk_header = &chunk_headers[shard_index];
```

Comment:
In `Chunks` struct, they say that `Index` trait is deprecated and recommends using `.get()` instead. Just to align with the new resharding implementattion.
```suggestion
        let chunk_header = &chunk_headers.get(shard_index);
```